Something is wrong, the test failure is:
```java
org.keycloak.admin.client.resource.RealmResource$$QuarkusRestClientInterface.testSMTPConnection(java.lang.String) 
2022-08-24T20:16:07.9059051Z Parsing context:
2022-08-24T20:16:07.9059562Z    at org.keycloak.admin.client.resource.RealmResource$$QuarkusRestClientInterface.testSMTPConnection(Unknown Source)
2022-08-24T20:16:07.9060272Z    at org.keycloak.admin.client.resource.RealmResource.testSMTPConnection(RealmResource.java)
2022-08-24T20:16:07.9066246Z 
2022-08-24T20:16:07.9074394Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:152)
2022-08-24T20:16:07.9081827Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:104)
2022-08-24T20:16:07.9088942Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureFlowsGraphCreated(MethodTypeFlow.java:83)
2022-08-24T20:16:07.9096037Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.getOrCreateMethodFlowsGraph(MethodTypeFlow.java:65)
2022-08-24T20:16:07.9103563Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.DefaultVirtualInvokeTypeFlow.onObservedUpdate(DefaultVirtualInvokeTypeFlow.java:109)
2022-08-24T20:16:07.9110307Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:558)
2022-08-24T20:16:07.9117291Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.PointsToAnalysis$1.run(PointsToAnalysis.java:635)
2022-08-24T20:16:07.9129605Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:193)
2022-08-24T20:16:07.9131000Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:177)
2022-08-24T20:16:07.9131639Z 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
2022-08-24T20:16:07.9132129Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
2022-08-24T20:16:07.9176437Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
2022-08-24T20:16:07.9177384Z ------------------------------------------------------------------------------------------------------------------------
2022-08-24T20:16:07.9216054Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
2022-08-24T20:16:07.9216626Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
2022-08-24T20:16:07.9217161Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
2022-08-24T20:16:07.9218630Z Caused by: org.graalvm.compiler.java.BytecodeParser$BytecodeParserError: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: java.lang.String.name(). This error is reported at image build time because class org.keycloak.admin.client.resource.RealmResource$$QuarkusRestClientInterface is registered for linking at image build time by command line
2022-08-24T20:16:07.9219744Z 	at parsing org.keycloak.admin.client.resource.RealmResource$$QuarkusRestClientInterface.testSMTPConnection(Unknown Source)
2022-08-24T20:16:07.9220415Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.throwParserError(BytecodeParser.java:2506)
2022-08-24T20:16:07.9221418Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.throwParserError(SharedGraphBuilderPhase.java:105)
2022-08-24T20:16:07.9222247Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3367)
2022-08-24T20:16:07.9222982Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.handleBytecodeBlock(BytecodeParser.java:3319)
2022-08-24T20:16:07.9223662Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3164)
2022-08-24T20:16:07.9224273Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1138)
2022-08-24T20:16:07.9224913Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1030)
2022-08-24T20:16:07.9225569Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
2022-08-24T20:16:07.9226266Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:79)
2022-08-24T20:16:07.9227003Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
2022-08-24T20:16:07.9227635Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:261)
2022-08-24T20:16:07.9228142Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
2022-08-24T20:16:07.9228623Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
2022-08-24T20:16:07.9229233Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:135)
2022-08-24T20:16:07.9229970Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:685)
2022-08-24T20:16:07.9230684Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:168)
2022-08-24T20:16:07.9231392Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:343)
2022-08-24T20:16:07.9232087Z 	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.MethodTypeFlow.createFlowsGraph(MethodTypeFlow.java:93)
2022-08-24T20:16:07.9232527Z 	... 13 more
2022-08-24T20:16:07.9233573Z Caused by: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: java.lang.String.name(). This error is reported at image build time because class org.keycloak.admin.client.resource.RealmResource$$QuarkusRestClientInterface is registered for linking at image build time by command line
2022-08-24T20:16:07.9234830Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.reportUnresolvedElement(SharedGraphBuilderPhase.java:298)
2022-08-24T20:16:07.9235778Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedMethod(SharedGraphBuilderPhase.java:288)
2022-08-24T20:16:07.9236968Z 	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.phases.SharedGraphBuilderPhase$SharedBytecodeParser.handleUnresolvedInvoke(SharedGraphBuilderPhase.java:244)
2022-08-24T20:16:07.9237916Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeVirtual(BytecodeParser.java:1709)
2022-08-24T20:16:07.9238599Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5222)
2022-08-24T20:16:07.9239306Z 	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3359)
```

--- a/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
+++ b/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
@@ -31,6 +31,7 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletionStage;
@@ -2532,15 +2533,36 @@ public class JaxrsClientReactiveProcessor {
                 notNullValue.load(methodIndex));
         ResultHandle isString = notNullValue.instanceOf(convertedFormParam, String.class);
         BranchResult isStringBranch = notNullValue.ifTrue(isString);
-        isStringBranch.falseBranch().throwException(IllegalStateException.class,
-                "Form parameter '" + paramName
-                        + "' could not be converted to 'String' for REST Client interface '"
-                        + restClientInterfaceClassName + "'. A proper implementation of '"
-                        + ParamConverter.class.getName() + "' needs to be returned by a '"
-                        + ParamConverterProvider.class.getName()
-                        + "' that is registered with the client via the @RegisterProvider annotation on the REST Client interface.");
         isStringBranch.trueBranch().invokeInterfaceMethod(MULTIVALUED_MAP_ADD, formParams,
                 notNullValue.load(paramName), convertedFormParam);
+
+        // if the converted value is not a string, then:
+        // - if it's a primitive type, use the valueOf() method.
+        if (EndpointIndexer.primitiveTypes.containsKey(parameterType)) {
+            ResultHandle convertedFormParamAsString = isStringBranch.falseBranch().invokeStaticMethod(
+                    MethodDescriptor.ofMethod(Objects.class, "toString", String.class, Object.class),
+                    convertedFormParam);
+            isStringBranch.falseBranch().invokeInterfaceMethod(MULTIVALUED_MAP_ADD, formParams,
+                    notNullValue.load(paramName), convertedFormParamAsString);
+        } else {
+            // - if it's an enum, use the name() method.
+            ResultHandle isEnum = isStringBranch.falseBranch().instanceOf(convertedFormParam, Enum.class);
+            BranchResult isEnumBranch = isStringBranch.falseBranch().ifTrue(isEnum);
+            ResultHandle enumAsString = isEnumBranch.trueBranch().invokeVirtualMethod(
+                    MethodDescriptor.ofMethod(Enum.class, "name", String.class),
+                    isEnumBranch.trueBranch().checkCast(convertedFormParam, Enum.class));
+            isEnumBranch.trueBranch().invokeInterfaceMethod(MULTIVALUED_MAP_ADD, formParams,
+                    notNullValue.load(paramName), enumAsString);
+
+            // - Otherwise, return exception
+            isEnumBranch.falseBranch().throwException(IllegalStateException.class,
+                    "Form parameter '" + paramName
+                            + "' could not be converted to 'String' for REST Client interface '"
+                            + restClientInterfaceClassName + "'. A proper implementation of '"
+                            + ParamConverter.class.getName() + "' needs to be returned by a '"
+                            + ParamConverterProvider.class.getName()
+                            + "' that is registered with the client via the @RegisterProvider annotation on the REST Client interface.");
+        }
     }
 
     private void addCookieParam(BytecodeCreator invoBuilderEnricher, AssignableResultHandle invocationBuilder,

--- a/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/FormParamTest.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/FormParamTest.java
@@ -24,7 +24,7 @@ public class FormParamTest {
     @RegisterExtension
     static final QuarkusUnitTest TEST = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
-                    .addClasses(FormClient.class, SubFormClient.class, Resource.class));
+                    .addClasses(FormClient.class, SubFormClient.class, Resource.class, Mode.class));
 
     @TestHTTPResource
     URI baseUri;
@@ -50,6 +50,13 @@ public class FormParamTest {
         assertThat(result).isEqualTo("sub rootParam1:par1,rootParam2:par 2,subParam1:spar1,subParam2:spar 2");
     }
 
+    @Test
+    void shouldSupportParsingDifferentTypes() {
+        FormClient formClient = RestClientBuilder.newBuilder().baseUri(baseUri).build(FormClient.class);
+        String result = formClient.withTypes("a", 1, 2, Mode.On);
+        assertThat(result).isEqualTo("root text:a,number:1,wrapNumber:2,mode:On");
+    }
+
     public interface FormClient {
         @POST
         @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
@@ -58,6 +65,13 @@ public class FormParamTest {
 
         @Path("/sub")
         SubFormClient subForm(@FormParam("rootParam1") String formParam1, @FormParam("rootParam2") String formParam2);
+
+        @POST
+        @Path("/types")
+        @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+        @Produces(MediaType.TEXT_PLAIN)
+        String withTypes(@FormParam("text") String text, @FormParam("number") int number,
+                @FormParam("wrapNumber") Integer wrapNumber, @FormParam("mode") Mode mode);
     }
 
     public interface SubFormClient {

--- a/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/Mode.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/Mode.java
@@ -1,0 +1,6 @@
+package io.quarkus.rest.client.reactive.form;
+
+public enum Mode {
+    Off,
+    On
+}

--- a/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/Resource.java
+++ b/extensions/resteasy-reactive/rest-client-reactive/deployment/src/test/java/io/quarkus/rest/client/reactive/form/Resource.java
@@ -30,4 +30,12 @@ public class Resource {
         return Response.ok(String.format("sub rootParam1:%s,rootParam2:%s,subParam1:%s,subParam2:%s",
                 rootParam1, rootParam2, subParam1, subParam2)).build();
     }
+
+    @POST
+    @Path("/types")
+    public Response getWithTypes(@RestForm String text, @RestForm int number, @RestForm Integer wrapNumber,
+            @RestForm Mode mode) {
+        return Response.ok(String.format("root text:%s,number:%s,wrapNumber:%s,mode:%s", text, number, wrapNumber, mode))
+                .build();
+    }
 }

